include/swift/AST/DiagnosticsSema.def

```diff
@@ -4605,10 +4605,11 @@ NOTE(duplicated_key_declared_here, none,
 // Generic specializations
 ERROR(cannot_explicitly_specialize_generic_function,none,
       "cannot explicitly specialize a generic function", ())
-ERROR(not_a_generic_definition,none,
-      "cannot specialize a non-generic definition", ())
 ERROR(not_a_generic_type,none,
       "cannot specialize non-generic type %0", (Type))
+WARNING(cannot_specialize_macro,none,
+      "specializing a macro with type parameters is deprecated; "
```

Comment:
I don't think we need this message because we can specialize a macro if its type is generic but they are also (mistakenly) allowed to be specialized even if it isn't, that's why we need a downgrade, regular diagnostic as a warning seems to do the job pretty well for that maintaining source compatibility in the process.

Reply:
The diagnosis would be `warning: cannot specialize non-generic type '() -> ()'`, which seems a very odd message for this situation. Maybe just `warning: cannot specialize a non-generic macro`?

Also, the issue with the penny-bot source compatibility failure is that the macro definition specifically expects to see the generic specialization (it looks for it in the generic context in the implementation of the macro and uses it). So this isn't a situation where it is only an odd addition at the use site, it's something where the macro author and all of the uses of the macro will have to coordinate to change to using a macro argument for that type string instead.

E.g. Until the upstream library macro author (DiscordBM) changes their macro implementation the warning will be unfixable downstream.

Just wanted to be sure you were aware of the context - happy to change back to a warning downgrade of the existing message if you still think that's best.

Reply:
Yeah, I'm aware of the consequences here, this something for the project to address and their clients to update. One note is that I don't know why the diagnostic is using `() -> ()`  instead of type of the macro though, I think we should fix that.

Reply:
If you want to look at the macro implementation. Specifically the generic argument is required by lines 39-43: https://github.com/DiscordBM/DiscordBM/blob/main/Macros/UnstableEnumMacro/UnstableEnum.swift

Reply:
Ultimately the type comes from the MacroDecl `getInterfaceType()`, which is constructed in `InterfaceTypeRequest::evaluate()` 2678-2693. The macro has an empty (but non-NULL) parameterList and a void `resultType`, so this returns `() -> ()` on line 2693

Reply:
In any case: done!

Reply:
Left a comment about this in review - https://github.com/swiftlang/swift/pull/75526#discussion_r1697705120